pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java

```diff
@@ -378,80 +378,95 @@ CompletableFuture<MessageId> internalSendAsync(Message<?> message) {
+                stats.incrementSendFailed();
                 try {
-                    if (e != null) {
-                        latencyHistogram.recordFailure(latencyNanos);
-                        stats.incrementSendFailed();
-                        onSendAcknowledgement(interceptorMessage, null, e);
-                        future.completeExceptionally(e);
-                    } else {
-                        latencyHistogram.recordSuccess(latencyNanos);
-                        publishedBytesCounter.add(msgSize);
-                        onSendAcknowledgement(interceptorMessage, interceptorMessage.getMessageId(), null);
-                        future.complete(interceptorMessage.getMessageId());
-                        stats.incrementNumAcksReceived(latencyNanos);
-                    }
+                    onSendAcknowledgement(msg, null, e);
+                    sendCallback.getFuture().completeExceptionally(e);
                 } finally {
-                    interceptorMessage.getDataBuffer().release();
+                    if (payload == null) {
+                        log.error("[{}] [{}] Payload is null when calling a failed onSendComplete, which is not"
+                                + " expected.", topic, producerName);
+                        return;
+                    }
+                    ReferenceCountUtil.safeRelease(payload);
                 }
-
-                while (nextCallback != null) {
-                    SendCallback sendCallback = nextCallback;
-                    MessageImpl<?> msg = nextMsg;
-                    // Retain the buffer used by interceptors callback to get message. Buffer will release after
-                    // complete interceptors.
-                    try {
-                        msg.getDataBuffer().retain();
-                        if (e != null) {
-                            stats.incrementSendFailed();
-                            onSendAcknowledgement(msg, null, e);
-                            sendCallback.getFuture().completeExceptionally(e);
-                        } else {
-                            onSendAcknowledgement(msg, msg.getMessageId(), null);
-                            sendCallback.getFuture().complete(msg.getMessageId());
-                            stats.incrementNumAcksReceived(System.nanoTime() - createdAt);
-                        }
-                        nextMsg = nextCallback.getNextMessage();
-                        nextCallback = nextCallback.getNextSendCallback();
-                    } finally {
-                        msg.getDataBuffer().release();
+            } else {
+                latencyHistogram.recordSuccess(latencyNanos);
+                publishedBytesCounter.add(msgSize);
+                stats.incrementNumAcksReceived(latencyNanos);
+                try {
+                    onSendAcknowledgement(msg, msg.getMessageId(), null);
+                    sendCallback.getFuture().complete(msg.getMessageId());
```

Comment:
We should call `ReferenceCountUtil.safeRelease(payload);` before `sendCallback.getFuture().complete(msg.getMessageId());`

The buffer is retained by the Pulsar Client itself. So we should release it before callback the future.

Reply:
Done